_source/_posts/2021-08-09-reactive-java.md

```diff
@@ -0,0 +1,791 @@
+```java
+package com.okta.developer.reactive.controller;
+
+public class SecureRandom {
+
+    private String value;
+
+    public String getValue() {
+        return value;
+    }
+
+    public void setValue(String value) {
+        this.value = value;
+    }
+
+    public SecureRandom(Integer value){
+        this.value = value.toString();
+    }
+}
+```
+Create the `SecureRandomController` class:
+
+```java
+package com.okta.developer.reactive.controller;
+
+import com.okta.developer.reactive.service.SecureRandomService;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.web.bind.annotation.GetMapping;
+import org.springframework.web.bind.annotation.RestController;
+import reactor.core.publisher.Mono;
+
+@RestController
+public class SecureRandomController {
+
+    @Autowired
+    private SecureRandomService secureRandomService;
+
+
+    @GetMapping("/random")
+    public Mono<SecureRandom> getSecureRandom(){
+        return secureRandomService.getRandomInt().map(i -> new SecureRandom(i));
+    }
+}
+```
+
+Add `spring-security-test` dependency to the pom:
+
+```xml
+<dependency>
+    <groupId>org.springframework.security</groupId>
+    <artifactId>spring-security-test</artifactId>
+    <version>5.4.5</version>
+    <scope>test</scope>
+</dependency>
+```
+
+
+Add a `SecureRandomControllerTest`:
+```java
+package com.okta.developer.reactive.controller;
+
+import org.junit.jupiter.api.Test;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.boot.test.autoconfigure.web.reactive.AutoConfigureWebTestClient;
+import org.springframework.boot.test.context.SpringBootTest;
+import org.springframework.test.context.ActiveProfiles;
+import org.springframework.test.web.reactive.server.WebTestClient;
+
+import static org.springframework.security.test.web.reactive.server.SecurityMockServerConfigurers.mockOidcLogin;
+
+@SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.RANDOM_PORT)
+@AutoConfigureWebTestClient
+@ActiveProfiles("test")
+public class SecureRandomControllerTest {
+
+    @Autowired
+    private WebTestClient webTestClient;
+
+
+    @Test
+    public void testGetSecureRandom() {
+
+
+        webTestClient.mutateWith(mockOidcLogin())
+            .get()
+            .uri("/random")
+            .exchange()
+            .expectStatus().isOk();
+    }
+}
+```
+
+
+Run with:
+```shell
+./mvnw test -Dtest=SecureRandomControllerTest
+```
+The test should pass, but **how can you make sure the REST call won't freeze up the service's event loop?** With [BlockHound](https://github.com/reactor/BlockHound), a Java agent to detect blocking calls from non-blocking threads. Blockhound has built-in integration with Project Reactor and supports the JUnit platform.
+
+Add the Blockhound dependency to the `pom.xml`:
+
+```xml
+<dependency>
+    <groupId>io.projectreactor.tools</groupId>
+    <artifactId>blockhound-junit-platform</artifactId>
+    <version>RELEASE</version>
```

Comment:
```suggestion
    <version>1.0.6.RELEASE</version>
```

nit: using the `RELEASE` keyword is a bad Maven practice

Reply:
Fixed